dotnet/src/Microsoft.Agents.AI.CosmosNoSql/CosmosChatMessageStore.cs

```diff
@@ -0,0 +1,666 @@
+    /// <param name="databaseId">The identifier of the Cosmos DB database.</param>
+    /// <param name="containerId">The identifier of the Cosmos DB container.</param>
+    /// <param name="jsonSerializerOptions">Optional settings for customizing the JSON deserialization process.</param>
+    /// <exception cref="ArgumentNullException">Thrown when <paramref name="cosmosClient"/> is null.</exception>
+    /// <exception cref="ArgumentException">Thrown when the serialized state cannot be deserialized.</exception>
+    public CosmosChatMessageStore(CosmosClient cosmosClient, JsonElement serializedStoreState, string databaseId, string containerId, JsonSerializerOptions? jsonSerializerOptions = null)
+    {
+        this._cosmosClient = cosmosClient ?? throw new ArgumentNullException(nameof(cosmosClient));
+        this.DatabaseId = Throw.IfNullOrWhitespace(databaseId);
+        this.ContainerId = Throw.IfNullOrWhitespace(containerId);
+        this._container = this._cosmosClient.GetContainer(databaseId, containerId);
+        this._ownsClient = false;
+
+        if (serializedStoreState.ValueKind is JsonValueKind.Object)
+        {
+            var state = JsonSerializer.Deserialize<StoreState>(serializedStoreState, jsonSerializerOptions);
+            if (state?.ConversationIdentifier is { } conversationId)
+            {
+                this.ConversationId = conversationId;
+
+                // Initialize hierarchical partitioning if available in state
+                this._tenantId = state.TenantId;
+                this._userId = state.UserId;
+                this._useHierarchicalPartitioning = state.UseHierarchicalPartitioning;
+
+                this._partitionKey = (this._useHierarchicalPartitioning && this._tenantId != null && this._userId != null)
+                    ? new PartitionKeyBuilder()
+                        .Add(this._tenantId)
+                        .Add(this._userId)
+                        .Add(conversationId)
+                        .Build()
+                    : new PartitionKey(conversationId);
+
+                return;
+            }
+        }
+
+        throw new ArgumentException("Invalid serialized state", nameof(serializedStoreState));
+    }
+
+    /// <inheritdoc />
+    public override async Task<IEnumerable<ChatMessage>> GetMessagesAsync(CancellationToken cancellationToken = default)
+    {
+#pragma warning disable CA1513 // Use ObjectDisposedException.ThrowIf - not available on all target frameworks
+        if (this._disposed)
+        {
+            throw new ObjectDisposedException(this.GetType().FullName);
+        }
+#pragma warning restore CA1513
+
+        // Fetch most recent messages in descending order when limit is set, then reverse to ascending
+        var orderDirection = this.MaxMessagesToRetrieve.HasValue ? "DESC" : "ASC";
+        var query = new QueryDefinition($"SELECT * FROM c WHERE c.conversationId = @conversationId AND c.type = @type ORDER BY c.timestamp {orderDirection}")
```

Comment:
Is the semantics always get TOP X ORDER_BY with no continutions?

Reply:
Good shout, we should add continuations, will add.